File: apps/api/app/schemas/skill_decay.py
"""
PathForge — Skill Decay & Growth Tracker Schemas
===================================================
Pydantic request/response models for the Skill Decay & Growth Tracker API.
All schemas use strict typing with no `Any` fields.
"""

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

# ── Skill Freshness ────────────────────────────────────────────


class SkillFreshnessResponse(BaseModel):
    """Single skill freshness score from the Skill Half-Life Engine™."""

    id: uuid.UUID
    skill_name: str
    category: str
    last_active_date: str | None = None
    freshness_score: float = Field(ge=0.0, le=100.0)
    half_life_days: int = Field(ge=1)
    decay_rate: str
    days_since_active: int = Field(ge=0)
    refresh_urgency: float = Field(ge=0.0, le=1.0)
    analysis_reasoning: str | None = None
    computed_at: datetime

    model_config = {"from_attributes": True}


# ── Market Demand ──────────────────────────────────────────────


class MarketDemandSnapshotResponse(BaseModel):
    """Single skill market demand snapshot from Market Demand Curves™."""

    id: uuid.UUID
    skill_name: str
    demand_score: float = Field(ge=0.0, le=100.0)
    demand_trend: str
    trend_confidence: float = Field(ge=0.0, le=1.0)
    job_posting_signal: dict[str, Any] | None = None
    industry_relevance: dict[str, Any] | None = None
    growth_projection_6m: float | None = None
    growth_projection_12m: float | None = None
    data_sources: dict[str, Any] | None = None
    snapshot_date: datetime

    model_config = {"from_attributes": True}


# ── Skill Velocity ─────────────────────────────────────────────


class SkillVelocityEntryResponse(BaseModel):
    """Single skill velocity from the Skill Velocity Map™."""

    id: uuid.UUID
    skill_name: str
    velocity_score: float = Field(ge=-100.0, le=100.0)
    velocity_direction: str
    freshness_component: float | None = None
    demand_component: float | None = None
    composite_health: float = Field(ge=0.0, le=100.0)
    acceleration: float | None = None
    reasoning: str | None = None
    computed_at: datetime

    model_config = {"from_attributes": True}


# ── Reskilling Pathways ───────────────────────────────────────


class ReskillingPathwayResponse(BaseModel):
    """Single personalized reskilling pathway recommendation."""

    id: uuid.UUID
    target_skill: str
    current_level: str
    target_level: str
    priority: str
    rationale: str | None = None
    estimated_effort_hours: int | None = None
    prerequisite_skills: dict[str, Any] | None = None
    learning_resources: dict[str, Any] | None = None
    career_impact: str | None = None
    freshness_gain: float | None = None
    demand_alignment: float | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Preferences ────────────────────────────────────────────────


class SkillDecayPreferenceResponse(BaseModel):
    """User notification preferences for skill decay tracking."""

    id: uuid.UUID
    tracking_enabled: bool = True
    notification_frequency: str = "weekly"
    decay_alert_threshold: float = Field(ge=0.0, le=100.0, default=40.0)
    focus_categories: dict[str, Any] | None = None
    excluded_skills: dict[str, Any] | None = None

    model_config = {"from_attributes": True}


class SkillDecayPreferenceUpdateRequest(BaseModel):
    """Update skill decay notification preferences (partial update)."""

    tracking_enabled: bool | None = None
    notification_frequency: str | None = None
    decay_alert_threshold: float | None = Field(
        default=None,
        ge=0.0,
        le=100.0,
    )
    focus_categories: dict[str, Any] | None = None
    excluded_skills: dict[str, Any] | None = None


# ── Request Schemas ───────────────────────────────────────────


class SkillRefreshRequest(BaseModel):
    """Request to manually refresh a skill's freshness score."""

    skill_name: str = Field(
        ...,
        min_length=1,
        max_length=255,
        description="Skill name to mark as refreshed",
    )


# ── Composite Responses ───────────────────────────────────────


class SkillDecayScanResponse(BaseModel):
    """Response after triggering a comprehensive skill decay scan."""

    status: str = "completed"
    skills_analyzed: int = 0
    freshness: list[SkillFreshnessResponse] = Field(default_factory=list)
    market_demand: list[MarketDemandSnapshotResponse] = Field(default_factory=list)
    velocity: list[SkillVelocityEntryResponse] = Field(default_factory=list)
    reskilling_pathways: list[ReskillingPathwayResponse] = Field(default_factory=list)


class SkillDecayDashboardResponse(BaseModel):
    """Full Skill Decay & Growth Tracker dashboard overview."""

    freshness: list[SkillFreshnessResponse] = Field(default_factory=list)
    freshness_summary: dict[str, Any] = Field(default_factory=dict)
    market_demand: list[MarketDemandSnapshotResponse] = Field(default_factory=list)
    velocity: list[SkillVelocityEntryResponse] = Field(default_factory=list)
    reskilling_pathways: list[ReskillingPathwayResponse] = Field(default_factory=list)
    preference: SkillDecayPreferenceResponse | None = None
    last_scan_at: datetime | None = None

File: packages/shared/src/types/api/skill-decay.ts
/**
 * PathForge — API Types: Skill Decay & Growth Tracker
 * =====================================================
 * Types for skill freshness, market demand, velocity, and reskilling.
 * Mirrors: apps/api/app/schemas/skill_decay.py
 */

// ── Skill Freshness ────────────────────────────────────────────

export interface SkillFreshnessResponse {
  readonly id: string;
  readonly skill_name: string;
  readonly category: string;
  readonly last_active_date: string | null;
  readonly freshness_score: number;
  readonly half_life_days: number;
  readonly decay_rate: string;
  readonly days_since_active: number;
  readonly refresh_urgency: number;
  readonly analysis_reasoning: string | null;
  readonly computed_at: string;
}

// ── Market Demand ──────────────────────────────────────────────

export interface MarketDemandSnapshotResponse {
  readonly id: string;
  readonly skill_name: string;
  readonly demand_score: number;
  readonly demand_trend: string;
  readonly trend_confidence: number;
  readonly job_posting_signal: Record<string, unknown> | null;
  readonly industry_relevance: Record<string, unknown> | null;
  readonly growth_projection_6m: number | null;
  readonly growth_projection_12m: number | null;
  readonly data_sources: Record<string, unknown> | null;
  readonly snapshot_date: string;
}

// ── Skill Velocity ─────────────────────────────────────────────

export interface SkillVelocityEntryResponse {
  readonly id: string;
  readonly skill_name: string;
  readonly velocity_score: number;
  readonly velocity_direction: string;
  readonly freshness_component: number | null;
  readonly demand_component: number | null;
  readonly composite_health: number;
  readonly acceleration: number | null;
  readonly reasoning: string | null;
  readonly computed_at: string;
}

// ── Reskilling Pathways ────────────────────────────────────────

export interface ReskillingPathwayResponse {
  readonly id: string;
  readonly target_skill: string;
  readonly current_level: string;
  readonly target_level: string;
  readonly priority: string;
  readonly rationale: string | null;
  readonly estimated_effort_hours: number | null;
  readonly prerequisite_skills: Record<string, unknown> | null;
  readonly learning_resources: Record<string, unknown> | null;
  readonly career_impact: string | null;
  readonly freshness_gain: number | null;
  readonly demand_alignment: number | null;
  readonly created_at: string;
}

// ── Preferences ────────────────────────────────────────────────

export interface SkillDecayPreferenceResponse {
  readonly id: string;
  readonly tracking_enabled: boolean;
  readonly notification_frequency: string;
  readonly decay_alert_threshold: number;
  readonly focus_categories: Record<string, unknown> | null;
  readonly excluded_skills: Record<string, unknown> | null;
}

export interface SkillDecayPreferenceUpdateRequest {
  readonly tracking_enabled?: boolean;
  readonly notification_frequency?: string;
  readonly decay_alert_threshold?: number;
  readonly focus_categories?: Record<string, unknown> | null;
  readonly excluded_skills?: Record<string, unknown> | null;
}

// ── Request Schemas ────────────────────────────────────────────

export interface SkillRefreshRequest {
  readonly skill_name: string;
}

// ── Composite Responses ────────────────────────────────────────

export interface SkillDecayScanResponse {
  readonly status: string;
  readonly skills_analyzed: number;
  readonly freshness: SkillFreshnessResponse[];
  readonly market_demand: MarketDemandSnapshotResponse[];
  readonly velocity: SkillVelocityEntryResponse[];
  readonly reskilling_pathways: ReskillingPathwayResponse[];
}

export interface SkillDecayDashboardResponse {
  readonly freshness: SkillFreshnessResponse[];
  readonly freshness_summary: Record<string, unknown>;
  readonly market_demand: MarketDemandSnapshotResponse[];
  readonly velocity: SkillVelocityEntryResponse[];
  readonly reskilling_pathways: ReskillingPathwayResponse[];
  readonly preference: SkillDecayPreferenceResponse | null;
  readonly last_scan_at: string | null;
}
